File: python/average-of-levels-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        elements = []
        s = [(root, 0)]
        while s:
            node, level = s.pop()
            if node == None:
                continue
            if len(elements) == level:
                elements.append([node.val, 1])
            else:
                elements[level][0] += node.val
                elements[level][1] += 1
            s.append((node.left, level + 1))
            s.append((node.right, level + 1))
        averages = [summatory/n for summatory, n in elements]
        return averages

File: rust/average-of-levels-in-binary-tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }

use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn average_of_levels(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<f64> {
        let mut averages = Vec::new();
        let mut dq = std::collections::VecDeque::new();
        dq.push_back(root);
        while !dq.is_empty() {
            let mut s = 0.0;
            let mut n = 0.0;
            for _ in 0..dq.len() {
                if let Some(Some(node)) = dq.pop_front() {
                    let node = node.borrow();
                    s += node.val as f64;
                    n += 1.0;
                    dq.push_back(node.left.clone());
                    dq.push_back(node.right.clone());
                }
            }
            if 0.0 < n {
                averages.push(s / n);
            }
        }
        return averages
    }
}
